Add KEP-167 'scheduler simulator operator' (#185)
kubernetes-sigs/kube-scheduler-simulator: keps/159-scheduler-simulator-operator/README.md

> # KEP-159: scheduler simulator operator

Reviewer: ```suggestion
# KEP-159: Scheduler Simulator Operator
```

## Summary

> A new custom resource to represent a simulator itself and a custom controller to manage it are introduced.

Reviewer 2: ```suggestion
A new custom resource to represent a simulator itself and a custom controller to manage it will be introduced.
```
?

## Motivation

With this, we can easily create/use the simulator as sandbox for the simulator by creating the Simulator custom resource. 

This feature also makes it easier to handle simulators from other CRD or controllers. 
They can gets the information for the simulator easily by accessint the `Simulator` resource.

### Goals

> - Users can define simulators with Simulator CRD.

Reviewer: ```suggestion
- Users can define simulators via `Simulator` CRs.
```

- The controller will manage(create, edit, delete) simulators.

### Non-Goals

Create new `Simulator` from Web UI. (may be implemented in the future, but out of scope of this proposal.)

## Proposal

### Implementation design details

#### Required prior knowledge

The current simulator is built in a single binary: the binary contains kube-apiserver, kube-scheduler and the simulator backend (which provides the web UI specific feature.)

> And, the simulator can be configured via environment variables

Reviewer 2: ```suggestion
And, the simulator can be configured via environment variables.
```

#### Simulator CRD

The CRD `Simulator` will be applied to kube-apiserver in user's cluster. (not to simulator's kube-apiserver.) 

It will create the Pod for the simulator in the same namespace.

```go
type Simulator struct {
  metav1.TypeMeta 
  metav1.ObjectMeta // namespaced.

  Spec SimulatorSpec
  Status SimulatorStatus
}

// SimulatorSpec has the spec for the simulator.

> // All fields can be updated (but a Pod may be re-created.)

Reviewer: Don't we need to handle existing(finished/running/pending) `Scenario` resources inside of the sumulator??  I think we should wait for users to clean up `Scenario` in the simulator. What do you think?

Author: Good question.

> we should wait for users to clean up Scenario in the simulator.

What did you mean by "clean up"? Let users delete all Scenario resources in the simulator? So, I don't want to let users do something for restarting the simulator Pod (since it's hard for users to realize they need to do that to kick restarting), but want the `Simulator` controller to do all the things for restarting the simulator Pod. 

How about this:
- implementing a way (endpoint on the simulator server?) that tells the Scenario controller that the Simulator controller wants to shut down the Pod.
  - when the Scenario controller moves FakeTime forward, it checks if the shutdown notification has come or not. If it has, the Scenario controller changes the Scenario's phase to "Paused" instead of continuing running the Scenario.
- setting `preStop` to notify shutdown to the Scenario controller by the above way.

Author: I think we can put this in "Future work", and until implementing them, the controller will behave like this: when the controller is started and finds the Scenario which phase is "Running", the controller just changes the status "Failed" with the `.status.message` like "the controller restarted while the Scenario was running".

Reviewer: > Let users delete all Scenario resources in the simulator?

Yeah, I thought so.

> How about this:

that sounds reasonable about handling the running Scenario. How about completed scenarios? If a Simulator CR (i.e. simulator pod) hosts multiple (say a lot of) completed Scenario CRs and what if a user updated or deleted the Simulator CR accidentally??  I think we should let users resurrect Scenario results in some way.

Author: >  How about completed scenarios?

We don't need to do anything. The etcd has the data for completed scenario.

> what if a user updated or deleted the Simulator CR accidentally??

Could you elaborate on that? The system should not resurrect what our users have updated/deleted.

Author: > I think we can put this in "Future work"

I'll write this in https://github.com/kubernetes-sigs/kube-scheduler-simulator/pull/186

Author: Added the note for this on https://github.com/kubernetes-sigs/kube-scheduler-simulator/pull/186

type SimulatorSpec struct {
  // KubeAPIServerPort indicates kube-apiserver’s URL which the simulator has internally.
  // (The each simulator has own kube-apiserver internally.)
  // This field's value will be added to a "simulator" container's env as "KUBE_API_PORT".

>   KubeAPIServerPort            int 

Reviewer 2: What is the `KubeAPIServerPort ` indicate? Port(int) or URL(string)?

Author: It's Port.

Author: Ooh, okay

  // - can be configured by the same environment variables. 
  // - (will add other requirements if any)
  Image                        string
  // Describes the pod that will be created.
  // This Pod should create the simulator.
  //
  // The system will add the container named "simulator" that: 
  // - run the simulator created by image specified in Image field,
  // - have Ports for kube-apiserver and the simulator endpoint,
  // - have environment variables to configure the simulator.
  // 
  // And if the EtcdURL field is empty, the system will add the container named "etcd" for kube-apiserver.
  // 
  // You can overwrite those container spec by defining containers that have the same name, "simulator" or "etcd".
  // 
  // When PodTemplate has the container(s) named "simulator" or "etcd" and the field is different from the container spec generated by other fields in SimulatorSpec,
  // container(s) definition in PodTemplate takes precedence.
  PodTemplate *corev1.PodTemplateSpec 
}

type SimulatorStatus struct {
  Phase SimulatorPhase

  // A human readable message indicating details about why the simulator is in this phase.
  // optional
  Message *string 

> }

Reviewer: I think it would be useful to kube-apiserver's url, kubeconfig, etc. information in the status.  Users will use them to create `Scenario` CRs in the simulator.

Author: sg 👍

  // Image should be the image of simulator.
  // You can use your customized simulator here, but the simulator container created by this image should follow these things:
  // - can be configured by the same environment variables. 
  // - (will add other requirements if any)
  Image                        string
  // Describes the pod that will be created.
  // This Pod should create the simulator.
  //
  // The system will add the container named "simulator" that: 
  // - run the simulator created by image specified in Image field,
  // - have Ports for kube-apiserver and the simulator endpoint,
  // - have environment variables to configure the simulator.
  // 
  // And if the EtcdURL field is empty, the system will add the container named "etcd" for kube-apiserver.
  // 
  // You can overwrite those container spec by defining containers that have the same name, "simulator" or "etcd".
  // 
  // When PodTemplate has the container(s) named "simulator" or "etcd" and the field is different from the container spec generated by other fields in SimulatorSpec,
  // container(s) definition in PodTemplate takes precedence.
  PodTemplate *corev1.PodTemplateSpec 

> }

Reviewer 2: What do you think about the `EXTERNAL_IMPORT_ENABLED` env for the external cluster import function?
Currently, the function is Beta one, and it will not be required to `Scenario-based simulation` and `scheduler-simulation`. But I think that it is important a part of the essential function for future.

Author: I think it's not required at least for now since users can set EXTERNAL_IMPORT_ENABLED in PodTemplate field if they want to use.

type SimulatorStatus struct {
  Phase SimulatorPhase

  // A human readable message indicating details about why the simulator is in this phase.
  // optional
  Message *string 

  // KubeAPIServerURL is the url to access the kube-apiserver in the simulator.

>   KubeAPIServerURL string

Reviewer: Let's add `KubeConfig` or `KubeConfigPath`.

Author: As we discussed offline, KubeConfig isn't needed since users can communicate with the kube-apiserver in simulator only with url.
